subprojects/vinumc/vinumc.c

```diff
@@ -29,19 +35,138 @@ void yyerror(char *s, ...) {
 
 extern FILE *yyin;
 
+enum flag_kind {
+	FLAG_BOOLEAN,
+	FLAG_ARGUMENT,
+};
+
+struct flag {
+	const char *name;
+	char short_name;
+	const char *help_desc;
+	const char *placeholder_name;
+	enum flag_kind kind;
+	union {
+		bool *boolean;
+		char **str;
+	} ref_as;
+};
+
+static struct flag* find_flag_by_short_name(struct flag *flags, size_t flags_len, char short_flag) {
+	for (size_t i = 0; i < flags_len; i++) {
+		struct flag *f = &flags[i];
+
+		if (f->short_name == short_flag)
+			return f;
+	}
+
+	return NULL;
+}
+
+static void parse_cmdline(const int argc, char **argv, struct ctx *ctx, struct flag *flags,
+			 size_t flags_len) {
+	char  *optstring = calloc(2 * (flags_len + 1), sizeof(*optstring));
+	struct option *longopts = calloc(flags_len + 1,  sizeof(*longopts));
+
+	size_t optstring_i = 0;
+	for (size_t i = 0; i < flags_len; i++) {
+		struct flag *f = &flags[i];
+
+		// TODO: support short/long only flags
+		assert(f->short_name != 0);
+		assert(f->name != 0);
+
+		optstring[optstring_i++] = f->short_name;
+
+		struct option opt = {.name = f->name, .val = f->short_name};
+			
+		switch (f->kind) {
+			case FLAG_ARGUMENT:
+				opt.has_arg = required_argument;
+				optstring[optstring_i++] = ':';
+				break;
+			case FLAG_BOOLEAN:
+				opt.has_arg = no_argument;
+				break;
+		}
+
+		longopts[i] = opt;
+	}
+
+	int opt;
+	while((opt = getopt_long(argc, argv, optstring, longopts, NULL)) != -1) {
+		struct flag *f = find_flag_by_short_name(flags, flags_len, opt);
+		assert(f != NULL);
+
+		switch (f->kind) {
+			case FLAG_ARGUMENT:;
+				size_t arg_len = strlen(optarg);
+				*f->ref_as.str = calloc(arg_len + 1, sizeof(char));
+				strcpy(*f->ref_as.str, optarg);
```

Comment:
I couldn't find a location were this memory is freed, I don't think it's a primary concern for now given that there are other, more relevant parts of the compiler that also leak memory.

Reply:
This is still not freed, but we can just add a TODO comment to fix it later, since we'll probably need to define ownership rules for pointers in the `ctx` struct since `ctx.input_path` directly receives a string from `argv` instead of a dynamically allocated string.

Reply:
Decided to solve these issues on this PR.